modules/core/plan/src/Entity/PlanRecord.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Drupal\plan\Entity;
+
+use Drupal\Core\Entity\ContentEntityBase;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Field\BaseFieldDefinition;
+
+/**
+ * Defines the Plan record relationship entity.
+ *
+ * This entity type can be used to create relationships between a plan and other
+ * record(s) along with additional metadata fields to describe the relationship.
+ *
+ * @ContentEntityType(
+ *   id = "plan_record",
+ *   label = @Translation("Plan record relationship"),
+ *   bundle_label = @Translation("Plan record relationship type"),
+ *   label_collection = @Translation("Plan record relationships"),
+ *   label_singular = @Translation("plan record relationship"),
+ *   label_plural = @Translation("plan record relationships"),
+ *   label_count = @PluralTranslation(
+ *     singular = "@count plan record relationship",
+ *     plural = "@count plan record relationships",
+ *   ),
+ *   handlers = {
+ *     "access" = "Drupal\plan\Access\PlanRecordAccess",
+ *     "form" = {
+ *        "edit" = "Drupal\Core\Entity\ContentEntityForm",
+ *      },
+ *     "route_provider" = {
+ *        "default" = "Drupal\Core\Entity\Routing\AdminHtmlRouteProvider",
+ *      },
+ *   },
+ *   base_table = "plan_record",
+ *   data_table = "plan_record_data",
+ *   entity_keys = {
+ *     "id" = "id",
+ *     "uuid" = "uuid",
+ *     "label" = "uuid",
+ *     "bundle" = "type",
+ *   },
+ *   bundle_entity_type = "plan_record_type",
+ *   common_reference_target = TRUE,
+ *   links = {
+ *      "edit-form" = "/plan/record/{plan_record}/edit",
+ *    },
+ * )
+ */
+class PlanRecord extends ContentEntityBase implements PlanRecordInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getBundleLabel() {
+    /** @var \Drupal\plan\Entity\PlanRecordTypeInterface $type */
+    $type = \Drupal::entityTypeManager()
```

Comment:
use of `\Drupal::` calls [should be avoided](https://github.com/mglaman/phpstan-drupal/blob/main/src/Rules/Drupal/GlobalDrupalDependencyInjectionRule.php#L75) in classes (with exception for static methods) instead dependency injection should be used, although in the case of entity types this isn't currently really possible https://www.drupal.org/project/drupal/issues/2142515
but EntityBase class which ContentEntityBase extends provides `entityTypeManager()` method, so maybe a good idea would be to use it here instead of directly calling `\Drupal::` so when DI capabilities for entity types will be added in core and the EntityBase::entityTypeManager() method will be updated to use it we could automatically take advantage of it without changing farmOS code
```suggestion
    $type = $this->entityTypeManager()
```

Reply:
Oh good point @wotnak. We actually do this in a few other entity types in farmOS core (most of the time we copy and paste code when creating new entity types). So I might open a separate PR to replace all of them together. Thanks for the heads up!

Reply:
I made a commit in a new `3.x-entitytypemanager` branch on top of this branch: https://github.com/mstenta/farmOS/commit/ffb1e8f7dcb4cdb4c730e1f992bbdea567a678ac

I'll open a PR to make that change once this is merged.

Reply:
https://github.com/farmOS/farmOS/pull/796